parsers/html.c

```diff
@@ -8,35 +8,434 @@
+	TOKEN_TAG_END,		/* >  */
+	TOKEN_TAG_END2,		/* /> */
+	TOKEN_EQUAL,
+	TOKEN_COMMENT,
+	TOKEN_OTHER
+} tokenType;
+
+typedef struct {
+	tokenType type;
+	vString *string;
+} tokenInfo;
+
+
+static int Lang_html;
+
+
+static void readTag (tokenInfo *token, vString *text);
+
+
 /*
-*   FUNCTION DEFINITIONS
-*/
+ * Perform the following regular expression matching to find the function name
+ * ^[ ]*function[ ]+([A-Za-z0-9_]+)[ ]*\\(
+ */
+static void findFunctionDecl (const vString *line)
+{
+	vString *fname;
+	const char *str = vStringValue (line);
+
+	if (vStringLength (line) < 11) /* strlen("function a(") */
+		return;
+
+	while (*str == ' ')
+		str++;
+
+	if (strncmp (str, "function", 8) != 0)
+		return;
+	str += 8;
+
+	if (*str != ' ')
+		return;
+
+	while (*str == ' ')
+		str++;
+
+	fname = vStringNew ();
+
+	while (isalnum (*str) || *str == '_')
+	{
+		vStringPut (fname, *str);
+		str++;
+	}
+
+	if (vStringLength (fname) > 0)
+	{
+		while (*str == ' ')
+			str++;
+
+		if (*str == '(')
+			makeSimpleTag (fname, HtmlKinds, K_FUNCTION);
+	}
+
+	vStringDelete (fname);
+}
+
+static void readTokenText (tokenInfo *const token, boolean collectText, boolean isScript)
+{
+	int c;
+	int lastC = 'X';  /* whatever, non-space character */
+	vString *line = NULL;
+
+	if (isScript)
+		line = vStringNew ();
+
+	vStringClear (token->string);
+
+getNextChar:
+
+	c = getcFromInputFile ();
+
+	switch (c)
+	{
+		case EOF:
+			token->type = TOKEN_EOF;
+			break;
+
+		case '<':
+		{
+			ungetcToInputFile (c);
+			token->type = TOKEN_TEXT;
+			break;
+		}
+
+		default:
+			if (collectText || isScript)
+			{
+				if (isScript && (c == '\r' || c == '\n'))
+				{
+					/* to get correct line number in findFunctionDecl() */
+					ungetcToInputFile (c);
+					findFunctionDecl (line);
+					getcFromInputFile ();
+					vStringClear (line);
+				}
+				else
+				{
+					if (c == '\r' || c == '\n' || c == '\t')
+						c = ' ';
+					if (c != ' ' || (c == ' ' && lastC != ' '))
+					{
+						if (collectText)
+							vStringPut (token->string, c);
+						if (isScript)
+							vStringPut (line, c);
+						lastC = c;
+					}
+				}
+			}
+
+			goto getNextChar;
+	}
+
+	if (isScript)
+	{
+		findFunctionDecl (line);
+		vStringDelete (line);
+	}
+}
+
+static void readToken (tokenInfo *const token, boolean skipComments)
+{
+	int c;
+
+	vStringClear (token->string);
+
+getNextChar:
+
+	c = getcFromInputFile ();
+	while (isspace (c))
+		c = getcFromInputFile ();
+
+	switch (c)
+	{
+		case EOF:
+			token->type = TOKEN_EOF;
+			break;
+
+		case '<':
+		{
+			int d = getcFromInputFile ();
+
+			if (d == '!')
+			{
+				d = getcFromInputFile ();
+				if (d == '-')
+				{
+					d = getcFromInputFile ();
+					if (d == '-')
+					{
+						int e = ' ';
+						int f = ' ';
+						do
+						{
+							d = e;
+							e = f;
+							f = getcFromInputFile ();
+						}
+						while (f != EOF && ! (d == '-' && e == '-' && f == '>'));
+
+						if (skipComments || f == EOF)
```

Comment:
this would skip the comment and return EOF when `skipComments == TRUE` and `f == EOF`, is that really wanted?  I'd rather see returning `TOKEN_COMMENT`, and let the next read return `TOKEN_EOF`.